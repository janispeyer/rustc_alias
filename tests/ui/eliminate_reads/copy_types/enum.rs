// compile-flags: -Zmir-emit-retag

#[derive(Copy, Clone)]
enum SomeEnum {
    A,
    B,
    C,
}

fn eliminate_reads(x: &mut SomeEnum, y: &mut SomeEnum) -> SomeEnum {
    *x = SomeEnum::B;
    *y = SomeEnum::C;
    return *x;
}

fn main() {
    let mut x = SomeEnum::A;
    let mut y = SomeEnum::A;
    eliminate_reads(&mut x, &mut y);
}
