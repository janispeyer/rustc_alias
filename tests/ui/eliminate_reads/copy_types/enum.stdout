# Analysing main
# Analysing eliminate_reads
# TopOfBorrowStack Analysis
bb0
[ 0] [_1, _2] <- Retag([fn entry] _1)
[ 1] [_1, _2] <- Retag([fn entry] _2)
[ 2] [_1, _2] <- StorageLive(_4)
[ 3] [_1, _2] <- Deinit(_4)
[ 4] [_1, _2] <- discriminant(_4) = 1
[ 5] [_1, _2] <- (*_1) = move _4
[ 6] [_1, _2] <- StorageDead(_4)
[ 7] [_1, _2] <- StorageLive(_5)
[ 8] [_1, _2] <- Deinit(_5)
[ 9] [_1, _2] <- discriminant(_5) = 2
[10] [_1, _2] <- (*_2) = move _5
[11] [_1, _2] <- StorageDead(_5)
[12] [_1, _2] <- _0 = (*_1)
[13] [] <- return

# FindImmutabilitySpans Analysis
bb0
[ 0] {} <- Retag([fn entry] _1)
[ 1] {} <- Retag([fn entry] _2)
[ 2] {} <- StorageLive(_4)
[ 3] {} <- Deinit(_4)
[ 4] {} <- discriminant(_4) = 1
[ 5] {_1: Span(bb0[5])} <- (*_1) = move _4
[ 6] {_1: Span(bb0[5])} <- StorageDead(_4)
[ 7] {_1: Span(bb0[5])} <- StorageLive(_5)
[ 8] {_1: Span(bb0[5])} <- Deinit(_5)
[ 9] {_1: Span(bb0[5])} <- discriminant(_5) = 2
[10] {_1: Span(bb0[5]), _2: Span(bb0[10])} <- (*_2) = move _5
[11] {_1: Span(bb0[5]), _2: Span(bb0[10])} <- StorageDead(_5)
[12] {_1: Span(bb0[5]), _2: Span(bb0[10])} <- _0 = (*_1)
[13] {_1: Top, _2: Top} <- return


# CFG Before
bb0
Retag([fn entry] _1)
Retag([fn entry] _2)
StorageLive(_4)
Deinit(_4)
discriminant(_4) = 1
(*_1) = move _4
StorageDead(_4)
StorageLive(_5)
Deinit(_5)
discriminant(_5) = 2
(*_2) = move _5
StorageDead(_5)
_0 = (*_1)
return

# CFG After
bb0
Retag([fn entry] _1)
Retag([fn entry] _2)
StorageLive(_4)
Deinit(_4)
discriminant(_4) = 1
_6 = move _4
(*_1) = _6
StorageDead(_4)
StorageLive(_5)
Deinit(_5)
discriminant(_5) = 2
(*_2) = move _5
StorageDead(_5)
_0 = _6
return

